File: Solutions/SymmetricTree.cpp
// Leetcode #101:
// Source: https://leetcode.com/problems/symmetric-tree/discuss/33054/Recursive-and-non-recursive-solutions-in-Java/31849

// Solution 1 (Recursive): O(n) time, O(n) space
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return isSymmetricHelper(root->left, root->right);
    }
  
    bool isSymmetricHelper(TreeNode* p, TreeNode* q) {
        // If both nodes are NULL, return true
        if (p == NULL && q == NULL) return true;
        
        // If only one node is NULL OR the nodes have different values, return false
        if (p == NULL || q == NULL || p->val != q->val) return false;
        
        // Recursively check if p's left subtree matches q's right subtree and vice versa
        return isSymmetricHelper(p->left, q->right) && isSymmetricHelper(p->right, q->left);
        
    }
};

// Solution 2 (Iterative w/ Stack): O(n) time, O(n) space

File: Solutions/SymmetricTree.py
# LC #101: https://leetcode.com/problems/symmetric-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Solution 1 (Recursive) | O(n) time, O(n) space
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def symmetric(n1, n2):
            if not n1 and not n2:
                # If both nodes are null, it is symmetric
                return True
            elif not n1 or not n2 or n1.val != n2.val:
                # If only one is null or if they have different values, it's not
                return False
            
            # Recursively check if n1's left subtree matches n2's righ subtree and vice versa
            return symmetric(n1.left, n2.right) and symmetric(n1.right, n2.left)
        
        return symmetric(root.left, root.right)
    
